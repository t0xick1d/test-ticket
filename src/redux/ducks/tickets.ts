import { createSlice } from '@reduxjs/toolkit';
import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react';

import { TicketsI } from '../../types/TicketsInterface';

export interface ticketStateI {
   data: TicketsI[];
}

const initialState: ticketStateI = {
   data: [],
};

export const ticketsSlice = createSlice({
   name: 'tickets',
   initialState,
   reducers: {},
});

export const ticketsApi = createApi({
   reducerPath: 'ticketsApi',
   baseQuery: fetchBaseQuery({
      baseUrl: 'https://keenethics-test-server.onrender.com/api/bicycle/',
   }),
   endpoints: (builder) => ({
      getSearchId: builder.query({
         query: () => `search`,
      }),
      getTickets: builder.query<TicketsI[], string>({
         query: (searchID) => `tickets/searchId=${searchID}`,
      }),
   }),
});

export const { useGetSearchIdQuery, useGetTicketsQuery } = ticketsApi;

// export const {} = ticketsSlice.actions;

export default ticketsSlice.reducer;
