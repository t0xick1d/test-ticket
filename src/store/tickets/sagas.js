import { call, put, all } from 'redux-saga/effects';

export const delay = (ms) => new Promise((res) => setTimeout(res, ms));

function* getTickets() {
   try {
      const searchId = yield call('https://front-test.beta.aviasales.ru/search');
      const ticktes = yield call(`https://front-test.beta.aviasales.ru/tickets?${searchId}`);
      yield put({ type: 'GET_TICKETS', ticktes });
   } catch (error) {
      yield put({ type: 'GET_TICKETS', error });
   }
}
export default function* rootSaga() {
   yield all([getTickets()]);
}  
