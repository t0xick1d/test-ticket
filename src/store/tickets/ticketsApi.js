import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react';

export const ticketsApi = createApi({
   reducerPath: 'ticketsApi',
   baseQuery: fetchBaseQuery({
      baseUrl: 'https://keenethics-test-server.onrender.com/api/bicycle/',
   }),
   endpoints: (builder) => ({
      getSearchId: builder.query({
         query: () => `search`,
      }),
      getTickets: builder.query({
         query: (searchID) => `tickets/searchId=${searchID}`,
      }),
   }),
});

export const { useGetSearchIdQuery, useGetTicketsQuery } = ticketsApi;
