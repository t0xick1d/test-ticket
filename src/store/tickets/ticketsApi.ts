import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react';
import { TicketsI } from '../../types/TicketsInterface';

export const ticketsApi = createApi({
   reducerPath: 'ticketsApi',
   baseQuery: fetchBaseQuery({
      baseUrl: 'https://keenethics-test-server.onrender.com/api/bicycle/',
   }),
   endpoints: (builder) => ({
      getSearchId: builder.query({
         query: () => `search`,
      }),
      getTickets: builder.query<TicketsI[], string>({
         query: (searchID) => `tickets/searchId=${searchID}`,
      }),
   }),
});

export const { useGetSearchIdQuery, useGetTicketsQuery } = ticketsApi;
