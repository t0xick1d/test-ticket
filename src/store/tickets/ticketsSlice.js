import { createSlice } from '@reduxjs/toolkit';
import { ticketsList } from '../../api';

export const ticketsSlice = createSlice({
   name: 'tickets',
   initialState: {
      data: ticketsList,
   },
   reducers: {},
});

export const { increment, decrement, incrementByAmount } = ticketsSlice.actions;

export default ticketsSlice.reducer;
