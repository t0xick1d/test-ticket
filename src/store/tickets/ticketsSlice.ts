import { createSlice } from '@reduxjs/toolkit';
import { TicketsI } from '../../types/TicketsInterface';

export interface ticketStateI {
   data: TicketsI[];
}

const initialState: ticketStateI = {
   data: [],
};

export const ticketsSlice = createSlice({
   name: 'tickets',
   initialState,
   reducers: {},
});

// export const {} = ticketsSlice.actions;

export default ticketsSlice.reducer;
