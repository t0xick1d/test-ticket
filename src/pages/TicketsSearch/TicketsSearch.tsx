import React, { ChangeEvent, useState } from 'react';
import { TicketsList, Filter, CountTransplants, Spinner } from '../../components';
import { useGetTicketsQuery, useGetSearchIdQuery } from '../../redux/ducks/tickets';

import style from './style.module.scss';

const TicketsSearch: React.FC = () => {
   const [alignment, setAlignment] = useState('cheap');
   const [transplants, setTransplants] = useState({
      all: true,
      noneTransplants: false,
      oneTransplants: false,
      twoTransplants: false,
      threeTransplants: false,
   });
   const searchid = useGetSearchIdQuery('');
   const ticket = useGetTicketsQuery('sdasdasdsa');
   const handleChange = (e: React.MouseEvent<HTMLElement>, newAlignment: string) => {
      setAlignment(newAlignment);
   };
 const transplantsChange = (event: ChangeEvent<HTMLInputElement>) => {
      const target = event.target as HTMLInputElement;
      setTransplants({
         ...transplants,
         [target.name]: target.checked,
      });
   };
   if (ticket.isLoading) {
      return <Spinner />;
   }
   if (ticket.error || searchid.error) {
      return <div>errorSearchID</div>;
   }
   return (
      <div className={style.App}>
         <section className={style.mainContainer}>
            <CountTransplants transplants={transplants} setTransplants={transplantsChange} />
            <div className={style.containerList}>
               <Filter alignment={alignment} handleChange={handleChange} />
               <TicketsList
                  list={ticket.data ? ticket.data : []}
                  alignment={alignment}
                  transplants={transplants}
               />
            </div>
         </section>
      </div>
   );
};

export default TicketsSearch;
