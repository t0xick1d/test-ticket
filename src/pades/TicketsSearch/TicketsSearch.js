import React, { useState } from 'react';
import { TicketsList, Filter, CountTransplants, Spinner } from '../../components';
import { useGetTicketsQuery, useGetSearchIdQuery } from '../../store/tickets/ticketsApi';

import style from './style.module.scss';

const TicketsSearch = () => {
   const [alignment, setAlignment] = useState('cheap');
   const [transplants, setTransplants] = useState({
      all: true,
      noneTransplants: false,
      oneTransplants: false,
      twoTransplants: false,
      threeTransplants: false,
   });
   const searchid = useGetSearchIdQuery();
   const ticket = useGetTicketsQuery('sdasdasdsa');

   const handleChange = (event, newAlignment) => {
      setAlignment(newAlignment);
   };
   const transplantsChange = (event) => {
      setTransplants({
         ...transplants,
         [event.target.name]: event.target.checked,
      });
   };
   if (ticket.isLoading) {
      return <Spinner />;
   }
   if (ticket.error || searchid.error) {
      return <div>errorSearchID</div>;
   }
   return (
      <div className={style.App}>
         <section className={style.mainContainer}>
            <CountTransplants transplants={transplants} setTransplants={transplantsChange} />
            <div className={style.containerList}>
               <Filter alignment={alignment} handleChange={handleChange} />
               <TicketsList list={ticket.data} />
            </div>
         </section>
      </div>
   );
};

export default TicketsSearch;
