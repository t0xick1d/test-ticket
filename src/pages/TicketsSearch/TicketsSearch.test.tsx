// TicketsSearch.test.tsx
import React from 'react';
import { render, screen } from '@testing-library/react';
import TicketsSearch from './TicketsSearch';

import { useGetTicketsQuery, useGetSearchIdQuery } from '../../redux/ducks/tickets';

jest.mock('../../redux/ducks/tickets', () => ({
   useGetTicketsQuery: jest.fn(),
   useGetSearchIdQuery: jest.fn(),
}));

jest.mock('../../components', () => ({
   TicketsList: () => <div>TicketsListMock</div>,
   Filter: () => <div>FilterMock</div>,
   CountTransplants: () => <div>CountTransplantsMock</div>,
   Spinner: () => <div>SpinnerMock</div>,
}));

describe('TicketsSearch', () => {
   beforeEach(() => {
      jest.clearAllMocks();
   });

   it('renders Spinner while loading', () => {
      (useGetTicketsQuery as jest.Mock).mockReturnValue({ isLoading: true });
      (useGetSearchIdQuery as jest.Mock).mockReturnValue({ error: null });

      render(<TicketsSearch />);

      expect(screen.getByText('SpinnerMock')).toBeInTheDocument();
   });

   it('renders error message on error', () => {
      (useGetTicketsQuery as jest.Mock).mockReturnValue({ isLoading: false, error: true });
      (useGetSearchIdQuery as jest.Mock).mockReturnValue({ error: false });

      render(<TicketsSearch />);

      expect(screen.getByText('errorSearchID')).toBeInTheDocument();
   });

   it('renders components with data', () => {
      (useGetTicketsQuery as jest.Mock).mockReturnValue({
         isLoading: false,
         error: null,
         data: [{ id: 1 }],
      });
      (useGetSearchIdQuery as jest.Mock).mockReturnValue({ error: null });

      render(<TicketsSearch />);

      expect(screen.getByText('CountTransplantsMock')).toBeInTheDocument();
      expect(screen.getByText('FilterMock')).toBeInTheDocument();
      expect(screen.getByText('TicketsListMock')).toBeInTheDocument();
   });

   it('matches snapshot', () => {
      (useGetTicketsQuery as jest.Mock).mockReturnValue({
         isLoading: false,
         error: null,
         data: [{ id: 1 }],
      });
      (useGetSearchIdQuery as jest.Mock).mockReturnValue({ error: null });

      const { asFragment } = render(<TicketsSearch />);
      expect(asFragment()).toMatchSnapshot();
   });
});
