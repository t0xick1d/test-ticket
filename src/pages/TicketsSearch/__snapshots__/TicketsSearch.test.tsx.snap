// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TicketsSearch matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <section
      class="mainContainer"
    >
      <div>
        CountTransplantsMock
      </div>
      <div
        class="containerList"
      >
        <div>
          FilterMock
        </div>
        <div>
          TicketsListMock
        </div>
      </div>
    </section>
  </div>
</DocumentFragment>
`;
